File: Java/src/main/java/OneLang/Generator/IGenerator/IGenerator.java
package OneLang.Generator.IGenerator;

import OneLang.Generator.GeneratedFile.GeneratedFile;
import OneLang.One.Ast.Types.Package;
import OneLang.One.ITransformer.ITransformer;
import OneLang.Generator.IGeneratorPlugin.IGeneratorPlugin;
import OneLang.One.Ast.Interfaces.IExpression;

import OneLang.One.ITransformer.ITransformer;
import OneLang.Generator.IGeneratorPlugin.IGeneratorPlugin;
import OneLang.One.Ast.Interfaces.IExpression;
import OneLang.Generator.GeneratedFile.GeneratedFile;
import OneLang.One.Ast.Types.Package;

public interface IGenerator {
    String getLangName();
    String getExtension();
    ITransformer[] getTransforms();
    void addPlugin(IGeneratorPlugin plugin);
    void addInclude(String include);
    String expr(IExpression expr);
    GeneratedFile[] generate(Package pkg);
}

File: PHP/src/OneLang/Generator/IGenerator.php
<?php

namespace OneLang\Generator\IGenerator;

use OneLang\Generator\GeneratedFile\GeneratedFile;
use OneLang\One\Ast\Types\Package;
use OneLang\One\ITransformer\ITransformer;
use OneLang\Generator\IGeneratorPlugin\IGeneratorPlugin;
use OneLang\One\Ast\Interfaces\IExpression;

interface IGenerator {
    function getLangName();
    
    function getExtension();
    
    function getTransforms();
    
    function addPlugin($plugin);
    
    function addInclude($include);
    
    function expr($expr);
    
    function generate($pkg);
}
